### PR TITLE
Destination Databricks: Implement refreshes

### PR DESCRIPTION
**Pull request by edgao**
closes https://github.com/airbytehq/airbyte-internal-issues/issues/8534. closes https://github.com/airbytehq/airbyte-internal-issues/issues/8835. closes https://github.com/airbytehq/airbyte-internal-issues/issues/8857#issuecomment-2259169235. Structurally identical to all the other refreshes PRs (e.g. https://github.com/airbytehq/airbyte/pull/38713).

❗ this PR also unpins the current version, i.e. we will release the connector publicly here.

As a refresher:
* pass the raw table suffix around to the relevant places
* DestinationHandler needs to fetch an InitialStatus for the temp raw table in addition to the real raw table
    * I did some minor code clarification in this file
* StorageOperation needs to implement some new methods
    * See the new integration test class for how we verify those new methods
    * some code shuffling in the StreamOperation to enable that integration test
* some new expectedrecords.jsonl files (CDK bump pulled in some new test cases)
* fixes a bug in `check` (wasn't deleting the test table). Added a test case for this.
    * (this only really matters for us - we had an older test table without generation_id, which caused check to fail. Real users wouldn't run into this... but it feels like best practice to clean up ourselves)
* fixes a bug in StreamOperation, where we would write an empty file, which then caused an error in the COPY command

I'll do a prerelease image + set up a sync in the perf test workspace, but in the meantime, this should be ready for review.

### CODE REVIEW

**Comment by edgao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/40692?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#40692** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/40692?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#40689** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/40689?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#40567** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/40567?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#40201** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/40201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#39008** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39008?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#39334](https://github.com/airbytehq/airbyte/pull/39334) <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39334?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#39141** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/39141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#38893** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/38893?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#38886** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/38886?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @edgao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt [hunk 1]
```diff
@@ -295,4 +303,6 @@ class DatabricksDestinationHandler(
     ) {
         // do Nothing
     }
+
+    fun query(query: String): List<JsonNode> = jdbcDatabase.queryJsons(query)
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt:307]: This method name could be more descriptive

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt [hunk 2]
```diff
@@ -79,35 +82,40 @@ class DatabricksDestinationHandler(
...
                 if (
                     existingTables.contains(namespace) &&
                         existingTables[namespace]?.contains(name) == true
                 ) {
+                    // The final table exists. Do some extra querying to find out what it looks
+                    // like.
+                    val isFinalTableSchemaMismatch =
+                        !isSchemaMatch(it, existingTables[namespace]?.get(name)!!)
+                    val isFinalTableEmpty = isFinalTableEmpty(it.id)
                     DestinationInitialStatus(
```
[the remark this answers is not in the document]
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt:100]: NVM, I just needed to refresh gradle in intelliJ, I wasn't looking at the local cdk.

[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt:100]: those were in a previous PR https://github.com/airbytehq/airbyte/pull/38622/files#diff-110c71ec8f8e76a0ec3e49f409fcb9517f4e2d6b213f81b93f4c71b22520d1fb

(this PR doesn't actually need local cdk, that's just me being lazy and not looking up the latest version number 😅 will fix before switching off draft mode)

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt (hunk 1 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt:307]: hm. how does `queryDestination(sql: String)` sound? anything more than that starts feeling more verbose than helpful

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/DatabricksDestination.kt [hunk 3]
```diff
@@ -97,8 +98,33 @@ class DatabricksDestination : BaseConnector(), Destination {
...
+                destinationHandler.execute(
+                    Sql.of(
+                        dropCheckTableStatement,
+                    ),
+                )
+            } catch (e: Exception) {
+                log.error(e) { "Failed to execute query $dropCheckTableStatement" }
+                return AirbyteConnectionStatus()
+                    .withStatus(AirbyteConnectionStatus.Status.FAILED)
+                    .withMessage("Failed to execute $dropCheckTableStatement: ${e.message}")
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/DatabricksDestination.kt:116]: I don't think we behave consistently across destinations wrt how we expose the actual error message here. Probably a thing we should think about for the new cdk

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt [hunk 4]
```diff
@@ -93,18 +92,9 @@ class DatabricksStorageOperation(
         )
     }
 
-    private fun prepareStagingTable(streamId: StreamId, destinationSyncMode: DestinationSyncMode) {
-        val rawSchema = streamId.rawNamespace
-        // TODO: Optimize by running SHOW SCHEMAS; rather than CREATE SCHEMA if not exists
-        destinationHandler.execute(sqlGenerator.createSchema(rawSchema))
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt:99]: the `create schema` call now happens via destinationhandler.createNamespaces, so remove it from here

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/DatabricksDestination.kt [hunk 5]
```diff
@@ -97,8 +98,33 @@ class DatabricksDestination : BaseConnector(), Destination {
                 syncId = 0
             )
 
+        // quick utility method to drop the airbyte_check_test_table table
+        // returns a connection status if there was an error, or null on success
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/DatabricksDestination.kt:102]: What's the purpose of that table?

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt (hunk 1 above)
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/jdbc/DatabricksDestinationHandler.kt:307]: Oh it's a verb. :)

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt [hunk 6]
```diff
@@ -116,11 +106,52 @@ class DatabricksStorageOperation(
         workspaceClient.files().createDirectory(stagingDirectory(streamId, database))
     }
 
-    override fun prepareStage(streamId: StreamId, destinationSyncMode: DestinationSyncMode) {
-        prepareStagingTable(streamId, destinationSyncMode)
+    override fun prepareStage(streamId: StreamId, suffix: String, replace: Boolean) {
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt:109]: Am I correct that this is never called with `replace = true` outside a test context?

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt [hunk 7]
```diff
@@ -116,11 +106,52 @@ class DatabricksStorageOperation(
...
+                """
+                INSERT INTO `$database`.`${streamId.rawNamespace}`.`${streamId.rawName}`
+                SELECT * FROM `$database`.`${streamId.rawNamespace}`.`${streamId.rawName}$suffix`
+                """.trimIndent(),
+                "DROP TABLE `$database`.`${streamId.rawNamespace}`.`${streamId.rawName}$suffix`",
+            )
+        )
+    }
+
+    override fun getStageGeneration(streamId: StreamId, suffix: String): Long? {
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt:143]: To satisfy my curiosity, do we pass any other suffixes through these functions besides `"_airbyte_tmp"` and `""`? Is this always really just `isTemp=true/false`?

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt (hunk 6 above)
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt:109]: Nvm. Via `AbstractStreamOperations::prepareStageForTruncate`

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/DatabricksDestination.kt (hunk 5 above)
Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/DatabricksDestination.kt:102]: Nvm. Does what it says. :)

File: airbyte-integrations/connectors/destination-databricks/metadata.yaml [hunk 8]
```diff
@@ -38,6 +36,7 @@ data:
     sl: 100
     ql: 100
   supportLevel: community
+  supportsRefreshes: true
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-databricks/metadata.yaml:39]: Is this the flag that will also cause the platform to send trace messages? Or is that something else?

**Review by johnny-schmidt — APPROVED** [review]
LGTM.
* the overridden methods do what they seem like they should
* the statements made about databricks in comments, etc are correct
* the net new behavior makes sense with my understanding of what supporting refreshes means
* the tests look thorough and correct but I didn't have time to go over them all in detail

**Review by gisripa — APPROVED** [review]

File: airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt (hunk 7 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-databricks/src/main/kotlin/io/airbyte/integrations/destination/databricks/operation/DatabricksStorageOperation.kt:143]: yep, that's correct for now. At some point I think we maybe should do generation-specific suffixes, but that's a whole different can of worms, and definitely not a consideration for the near future

File: airbyte-integrations/connectors/destination-databricks/metadata.yaml (hunk 8 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-databricks/metadata.yaml:39]: yep, that's all controlled by this flag

**Pull request merged**